### DISCUSSION TITLE
[2.5D Map] Reorganized (simplified for me) the 2.5D map code

### DISCUSSION THREAD

**Discussion by wolfgang-ch** [mytourbook/mytourbook#858] [General]
Reorganized (simplified for me) the 2.5D map code

* Fixed: Satellites images have been loaded even when the satellite layer was disabled
* Less flickering when selecting map provider/theme
* See issues in the commit comment in https://github.com/mytourbook/mytourbook/commit/d047c4b1cf235987d6a5d8d28414299a4051de05
* I don't see a difference between S3DB_Building and Building layer

@telemaxx After many code changes, can you check if your 2.5D map is still working as before by using this branch https://github.com/mytourbook/mytourbook/tree/ws-improvements

**Comment by wolfgang-ch** [comment]
> See issues in the commit comment in [d047c4b](https://github.com/mytourbook/mytourbook/commit/d047c4b1cf235987d6a5d8d28414299a4051de05)

I discovered that these issues also appear in the `main` branch

**Comment by telemaxx** [comment]
> I don't see a difference between S3DB_Building and Building layer

have you used mapsforge v5 maps from mapsforge?
only that maps contain s3db.

next week I am not at home,  I am back on Sunday. than I can do the tests. 

BTW: my new laptop from 2021  can not display buildings. this is a known problem from vtm with ryzen gpu.

but I can test in my VM, there the buildings normally ok.

**Comment by telemaxx** [comment]
Just tested latest main:
![grafik](https://user-images.githubusercontent.com/22706487/167458353-3c056da5-3971-45e9-91bd-74c7c30d9eea.png)

inside my Linux VM S3DB seems to be ok:
![grafik](https://user-images.githubusercontent.com/22706487/167458563-095170e0-bbde-421b-abba-7d02796212f3.png)

but i can not import GPX files in that version, dont know why.
![grafik](https://user-images.githubusercontent.com/22706487/167458913-95d0f1eb-1324-4410-b454-e1a2e09839f4.png)

Fit files are working.

**Reply by telemaxx** [reply]
ws-map25-direction-arrows did not run at the moment.

![grafik](https://user-images.githubusercontent.com/22706487/167460427-bccd2bf6-6ec1-410d-b729-89f8d0180a02.png)

i think i have to configure something...

**Reply by wolfgang-ch** [reply]
> Just tested latest main:

I've changed only the `ws-improvements` branch and NOT the `main` branch

>can you check ... by using this branch https://github.com/mytourbook/mytourbook/tree/ws-improvements

I don't want to merge these changes into `main` when I'm not sure if it works

Can you do the test again with the `ws-improvements` branch?

**Reply by wolfgang-ch** [reply]
> ws-map25-direction-arrows did not run at the moment.
> 
> ![grafik](https://user-images.githubusercontent.com/22706487/167460427-bccd2bf6-6ec1-410d-b729-89f8d0180a02.png)
> 
> i think i have to configure something...

The branch `ws-improvements` should not depend on **vtm-parent**. 

**vtm-parent** is the root project/plugin for all vtm projects which I can use to modify the original code and test it

![image](https://user-images.githubusercontent.com/1283445/167545958-5d37b173-1530-4090-9192-9e9bc22346b9.png)

**Reply by telemaxx** [reply]
> > Just tested latest main:
> 
> I've changed only the `ws-improvements` branch and NOT the `main` branch
> 
> > can you check ... by using this branch https://github.com/mytourbook/mytourbook/tree/ws-improvements
> 
> I don't want to merge these changes into `main` when I'm not sure if it works
> 
> Can you do the test again with the `ws-improvements` branch?

yes and that is running. but with my "old" arrows.

![grafik](https://user-images.githubusercontent.com/22706487/167677240-ba02655a-8ad9-4f89-99b4-f9909e20acc0.png)

**Reply by wolfgang-ch** [reply]
>yes and that is running. but with my "old" arrows.

I started the "new arrows" but it is so complicated with OpenGL (you have to understand how the math works, e.g. dot vector product, cross vector product and what you can do with it), that I started first the simple things

**Comment by telemaxx** [comment]
i just tried with my production version to import old GPX file.
same error.
seems to be a problem with date format.

java.text.ParseException: Unparseable date: "2010-06-02T17:10:23"

@wolfgang-ch  and @FJBDev can you try GPX import?

normaly i dont use GPX anymore.

**Reply by FJBDev** [reply]
GPX import works for me on `main` and 22.3.1 prod

![image](https://user-images.githubusercontent.com/25235665/167464961-f0bba0e4-55df-4ebd-aa48-935117f35149.png)

Can you attach the GPX file that causes you trouble ?

**Reply by telemaxx** [reply]
[13.10.Zwïesel_part_1.zip](https://github.com/mytourbook/mytourbook/files/8654440/13.10.Zwiesel_part_1.zip)

**Reply by telemaxx** [reply]
![grafik](https://user-images.githubusercontent.com/22706487/167476939-2ed07ad1-441c-4223-981c-39b48287b6d8.png)

**Reply by telemaxx** [reply]
not every gpx file is affected.

**Reply by FJBDev** [reply]
It seems that this file has the date times in a format not recognized by MT.
If you look at the `GPX_SAX_Handler`, then only [those time formats](https://github.com/mytourbook/mytourbook/blob/main/bundles/net.tourbook.device/src/net/tourbook/device/gpx/GPX_SAX_Handler.java#L230) are recognized.

In your file, I see this format, which I can understand why MT doesn't recognize because this format doesn't specify the timezone so MT has no reference to know the epoch time.

`<time>2011-10-13T12:09:21.42</time>`

I tried with previous MT versions like 20.12 and it fails too:
![image](https://user-images.githubusercontent.com/25235665/167507235-23ece405-34e9-496e-b485-934e3569e5b4.png)

**Comment by telemaxx** [comment]
Sunday we nearly hit the 20 degree

![grafik](https://user-images.githubusercontent.com/22706487/167462436-130b15fe-2d77-44af-8de8-fd76b360fb84.png)

**Comment by telemaxx** [comment]
for the map25 tests i started with a empty DB
than i imported an old(2010) GPX file, which fails.
seems so that a "Z" is missing at the end.

the newer ones i created with orux maps.
eg:
<time>2018-07-26T14:44:54Z</time>
this are OK

**Reply by wolfgang-ch** [reply]
The GPX importer can be extended to accept also date/time without time zone, which I did some weeks ago for another importer but cannot remember which it was.

**Comment by telemaxx** [comment]
the Conti Tower can fly:

![grafik](https://user-images.githubusercontent.com/22706487/167678928-ef857c98-5480-4c44-9e96-e09478c3e476.png)



![grafik](https://www.hannover.de/var/storage/images/_aliases/alias_1600xVariabel/media/01-data-neu/bilder/landeshauptstadt-hannover/politik/b%C3%BCrgerbeteiligung/wasserstadt/wasserstadt-limmer/11977324-1-ger-DE/Wasserstadt-Limmer.jpg)

Same Perspective:

![grafik](https://user-images.githubusercontent.com/22706487/167680008-b28fe60d-36f8-4f20-a371-56f3f04d7dcf.png)

**Reply by wolfgang-ch** [reply]
@telemaxx I don't can see the colored roofs and tried many maps and themes

Here https://mytourbook.sourceforge.io/mytourbook/index.php/documentation/preferences/pref-map25/pref-map25-installed-offline-map you write
>Search „Elevate 4“ then „manual download

but Elevation 4 cannot be downloaded anymore, only Elevation 5.

Can you provide links for the map and theme, with which the roofs can be displayed?

**Reply by telemaxx** [reply]
maps with roofs from here:
https://download.mapsforge.org/maps/v5/europe/

themes:
https://www.openandromaps.org/kartenlegende/elevation-hike-theme
ELV5 is correct, v4 is eol
the Theme from OAM is primary for oam maps, but can be used also for other maps. but hiking/cycling/biking is has only effect with oam maps. 

OAM maps > nice routes, no roofs
mapsforge maps > no routes, nice roofs.

there is a trick: put both map types in one folder, than you see routes and roofs. not perfect, but ok

**Reply by wolfgang-ch** [reply]
> the Conti Tower can fly:

Also the towers of a church in Passau

![image](https://user-images.githubusercontent.com/1283445/168608057-a186a8d3-d516-42b2-b82c-5b4c578a58af.png)

**Comment by wolfgang-ch** [comment]
Now I also see the buildings but it works only with min zoom level 17 

I increased the max tilt, it was 85°(2.5D map default is 65°) ....

![image](https://user-images.githubusercontent.com/1283445/168235731-aee3d6a4-fe75-426a-9619-fc4236355ff6.png)


...now it's 180° and you can see the buildings also from under the earth

![image](https://user-images.githubusercontent.com/1283445/168235946-de06e010-337a-4c27-8179-0b80ccc98105.png)

**Comment by wolfgang-ch** [comment]
I discovered, that when the S3DB Layer is not created, by commenting out the following lines and using an online map provider (I forgot to comment out the first line in my last commit, but this doesn't matter) ...

https://github.com/mytourbook/mytourbook/blob/8bf3b0ef150c6bc45c061f3212dcfb0f0f6f3e63/bundles/net.tourbook/src/net/tourbook/map25/Map25App.java#L559-L560

... then the buildings are not displayed, only the shadows (when enabled) and these errors occure

![image](https://user-images.githubusercontent.com/1283445/168483211-9281c9f6-196a-4144-b3c1-fe1b449a5809.png)

Even when the S3DB layer is not added to the visible layers, tile data are still loaded and provide building info. 
I'm struggeling since 2 days to find the reason for these errors (when I tried to simplify the building layers) now I found it :-)

With S3DB layer

![image](https://user-images.githubusercontent.com/1283445/168483475-adc49c5e-5ccc-4143-be8c-b96374bdc628.png)

Without S3DB layer

![image](https://user-images.githubusercontent.com/1283445/168483486-797e7994-8259-4595-8eb4-292d661637e8.png)

@telemaxx Are you aware of this behavior

**Reply by telemaxx** [reply]
I have fighting a lot with the building/s3db layers.

one importing thing was: when switching from online to offline or vice-versa then first all layers must be removed the the new source must set and then the layers must be recreated.

https://github.com/telemaxx/mytourbook/blob/main/bundles/net.tourbook/src/net/tourbook/map25/Map25App.java#L916

**Comment by wolfgang-ch** [comment]
Now only the S3DBLayer is used for buildings, for off- and online map's and I don't see any issues.

@telemaxx Can you test again the branch https://github.com/mytourbook/mytourbook/tree/ws-improvements ?

Shadow and min zoom level for buildings can now also be selected, each modification will replace the building layer

![image](https://user-images.githubusercontent.com/1283445/168759173-7a908d4c-f86b-44d4-b626-f83d3df5c824.png)

![image](https://user-images.githubusercontent.com/1283445/168760336-d691262e-1b2d-4424-bfa2-2df99eb98ce8.png)

**Reply by telemaxx** [reply]
> Now only the S3DBLayer is used for buildings, for off- and online map's and I don't see any issues.
> 
> @telemaxx Can you test again the branch https://github.com/mytourbook/mytourbook/tree/ws-improvements ?
> 

Yes of course, Friday I am back at home.

**Comment by telemaxx** [comment]
i have seen some news in main branch, so i started the tests here.

MAIN:
Online Mapilion maps looking great and is still very fast.
Offline maps are fast, but no buildings. after switching to mapilion there are also now buildings.

WS-improvements:
Online looks also good, new function "Show label before building" is also good.
Clicking on "Min zoom level -" changes value directly from 18 to 12 (Problem of linux or VM?)
normal buildings are missing then.
![grafik](https://user-images.githubusercontent.com/22706487/169533287-47e7e5e7-e2d4-44b4-81b3-f550bbfcde4d.png)
1st map is redrawn, then slider text changes to 12. that is a bit confusing.
clicking on + changes to 17 and buildings are ok than.
Maybe we should limit the range from 16?

my conti tower combined with sattelite
![grafik](https://user-images.githubusercontent.com/22706487/169535181-e76848c8-c0f2-4bd7-b744-f8e312369c2f.png)

combined V5 and V4 maps. so we se S3DB and colored routes (R7)
![grafik](https://user-images.githubusercontent.com/22706487/169535990-fb32ea16-1c4f-4404-bf9c-be9566dd122d.png)

**Reply by wolfgang-ch** [reply]
> Clicking on "Min zoom level -" changes value directly from 18 to 12 (Problem of linux or VM?)
> normal buildings are missing then.

When I test it with Ubuntu 20.04 in my VM then the +/- buttons for the zoom level is changing only one step after the other, however I already have seen in other spinner controls that it is not always changing one step after the other. I will change the page increment from 2 -> 1.

> Maybe we should limit the range from 16?

The tooltip displays the issues which can occure

![image](https://user-images.githubusercontent.com/1283445/169551708-bec528c9-c08f-4ed9-9966-5ae68bc6b571.png)

I will merge ws-improvements branch into main after it is working as before.

**Reply by wolfgang-ch** [reply]
> "Show label before building" is also good.

Sometimes you want to see the buildings better and sometimes the labels better, this is the reason why I implemented it.
